### PR TITLE
Add StarRocks connector

### PR DESCRIPTION
**Pull request by chenjian2664**
## Description
Fix https://github.com/trinodb/trino/issues/17329

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# StarRocks
* Add StarRocks connector. ({issue}`17329`)
```

### CODE REVIEW

File: plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java
Comment (Praveen2112) [inline plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java:1]: Since it supports MySQL protocol can we use `MysqlClient` here ?

File: plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksJdbcConfig.java
Comment (Praveen2112) [inline plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksJdbcConfig.java:21]: Can we add a test case for `isUrlValid`

File: plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksJdbcConfig.java
Comment (Praveen2112) [inline plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksJdbcConfig.java:28]: Can we use `assertFailsValidation` and specify the field here.

File: plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestingStarRocksContainer.java
Comment (Praveen2112) [inline plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestingStarRocksContainer.java:29]: Can we extract `DockerImageName` as a final variable

**Comment by chenjian2664** [comment]
@Praveen2112 Thank you so much for the review! I will mark the PR as the draft I don't have time now, I will open this PR after the 9th, Sorry about the mass.

File: plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksJdbcConfig.java
Comment (chenjian2664) [reply plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksJdbcConfig.java:34]: Can elaborate more on this?

File: plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksConnectorTest.java
Comment (chenjian2664) [reply plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksConnectorTest.java:202]: That would case code format validation fail

File: plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java
Comment (chenjian2664) [reply plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java:1]: Some mysql specify properties on office doc did not make clear whether they support or not. StarRocks also has some specified properties and DDL statement, from the system evolve persective I think we may not use`MysqlClient` directly, and Using `MysqlClient` also need depend on mysql plugin it's a little bit overkill.

File: plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksTableProperties.java
Comment (chenjian2664) [reply plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksTableProperties.java:59]: Yes, currently we only support `OLAP` engine type.

File: testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvMultinodeStarRocks.java
Comment (chenjian2664) [reply testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvMultinodeStarRocks.java:61]: After specified version, is it still need about this

File: plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksConnectorTest.java
Comment (chenjian2664) [reply plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksConnectorTest.java:249]: Currently not support unbounded varchar in StarRocks Connector

File: plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksJdbcConfig.java
Comment (chenjian2664) [reply plugin/trino-starrocks/src/test/java/io/trino/plugin/starrocks/TestStarRocksJdbcConfig.java:34]: Updated the test using `assertValidates`

**Comment by chenjian2664** [comment]
```
2023-09-29T15:52:32.981Z	WARN	main	io.trino.tests.product.launcher.env.Environment	+++++++++Container starrocks++++, id: 8f64dc15b505d332f9e15b37299a3c9f24ee62b25c94349b223f0f2afacb8966, start up time:26.02s, healthy: false
2023-09-29T15:52:32.996Z	WARN	main	io.trino.tests.product.launcher.env.Environment	Could not start environment 'multinode-star-rocks'
java.lang.RuntimeException: com.github.dockerjava.api.exception.NotFoundException: Status 404: {"message":"No such container: 8f64dc15b505d332f9e15b37299a3c9f24ee62b25c94349b223f0f2afacb8966"}

	at io.trino.tests.product.launcher.env.Environment.tryStart(Environment.java:177)
	at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:243)
	at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
	at dev.failsafe.internal.RetryPolicyExecutor.lambda$apply$0(RetryPolicyExecutor.java:74)
	at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:187)
	at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
	at dev.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:112)
	at io.trino.tests.product.launcher.env.Environment.start(Environment.java:130)
	at io.trino.tests.product.launcher.cli.TestRun$Execution.startEnvironment(TestRun.java:310)
	at io.trino.tests.product.launcher.cli.TestRun$Execution.tryExecuteTests(TestRun.java:255)
	at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:243)
	at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
	at dev.failsafe.internal.TimeoutExecutor.lambda$apply$1(TimeoutExecutor.java:93)
	at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:187)
	at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
	at dev.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:112)
	at io.trino.tests.product.launcher.cli.TestRun$Execution.call(TestRun.java:233)
	at io.trino.tests.product.launcher.cli.SuiteRun$Execution.runTest(SuiteRun.java:303)
	at io.trino.tests.product.launcher.cli.SuiteRun$Execution.executeSuiteTestRun(SuiteRun.java:281)
	at io.trino.tests.product.launcher.cli.SuiteRun$Execution.lambda$runSuites$2(SuiteRun.java:212)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Collections$2.tryAdvance(Collections.java:4853)
	at java.base/java.util.Collections$2.forEachRemaining(Collections.java:4861)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at io.trino.tests.product.launcher.cli.SuiteRun$Execution.lambda$runSuites$3(SuiteRun.java:213)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at io.trino.tests.product.launcher.cli.SuiteRun$Execution.runSuites(SuiteRun.java:199)
	at io.trino.tests.product.launcher.cli.SuiteRun$Execution.call(SuiteRun.java:185)
	at io.trino.tests.product.launcher.cli.SuiteRun$Execution.call(SuiteRun.java:136)
	at io.trino.tests.product.launcher.cli.LauncherCommand.call(LauncherCommand.java:61)
	at io.trino.tests.product.launcher.cli.LauncherCommand.call(LauncherCommand.java:29)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at io.trino.tests.product.launcher.cli.Launcher.execute(Launcher.java:77)
	at io.trino.tests.product.launcher.cli.Launcher.main(Launcher.java:66)
Caused by: com.github.dockerjava.api.exception.NotFoundException: Status 404: {"message":"No such container: 8f64dc15b505d332f9e15b37299a3c9f24ee62b25c94349b223f0f2afacb8966"}

	at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder.execute(DefaultInvocationBuilder.java:241)
	at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder.get(DefaultInvocationBuilder.java:202)
	at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder.get(DefaultInvocationBuilder.java:74)
	at org.testcontainers.shaded.com.github.dockerjava.core.exec.InspectContainerCmdExec.execute(InspectContainerCmdExec.java:31)
	at org.testcontainers.shaded.com.github.dockerjava.core.exec.InspectContainerCmdExec.execute(InspectContainerCmdExec.java:13)
	at org.testcontainers.shaded.com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21)
	at org.testcontainers.shaded.com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:33)
	at org.testcontainers.shaded.com.github.dockerjava.core.command.InspectContainerCmdImpl.exec(InspectContainerCmdImpl.java:51)
	at org.testcontainers.containers.ContainerState.getCurrentContainerInfo(ContainerState.java:129)
	at io.trino.tests.product.launcher.env.Environment.containerIsHealthy(Environment.java:338)
	at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
	at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:1002)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
	at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.allMatch(ReferencePipeline.java:637)
	at io.trino.tests.product.launcher.env.Environment.allContainersHealthy(Environment.java:327)
	at io.trino.tests.product.launcher.env.Environment.tryStart(Environment.java:167)
	... 44 more
```
cc @wendigo Do you know why encounter this error? The `Environment.containerIsHealthy` returns `false` for the `StarRocks` container, so I log the error, it shows the container Id exists, but fail on `getCurrentContainerInfo` due to above error

File: .github/workflows/ci.yml
Comment (hashhar) [inline .github/workflows/ci.yml:941]: nit: for consistency

```suggestion
            - suite-starrocks
```

File: plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java
Comment (hashhar) [inline plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java:142]:
```suggestion
    private static final LocalDate MIN_SUPPORTED_DATE = LocalDate.parse("0000-01-01");
    private static final LocalDate MAX_SUPPORTED_DATE = LocalDate.parse("9999-12-31");
```

File: plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java
Comment (hashhar) [inline plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java:271]: should engine type be quoted to prevent injection?

File: plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java
Comment (hashhar) [inline plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java:300]: Move this above `buildCreateTableProperties` (the order in which they are invoked) - easier to read/jump in code.

File: plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java
Comment (hashhar) [inline plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java:497]: why is table property called `distribution_desc` even though StarRocks itself calls it `distribute_key`?

**Review by hashhar — COMMENTED** [review]
skimmed, did not look at tests.

Looks in pretty good shape.

Sorry for the delay here.

File: plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java
Comment (chenjian2664) [reply plugin/trino-starrocks/src/main/java/io/trino/plugin/starrocks/StarRocksClient.java:497]: Yeah, seems StarRocks prefer `_desc` suffix to `_key` suffix, see [key_desc](https://docs.starrocks.io/en-us/2.5/sql-reference/sql-statements/data-definition/CREATE_TABLE#key_desc), [partition_desc](https://docs.starrocks.io/en-us/2.5/sql-reference/sql-statements/data-definition/CREATE_TABLE#partition_desc),etc.

**Comment by chenjian2664** [comment]
@mosabua Yes. I would greatly appreciate it if you/or someone could take a moment to review this pr

**Comment by mosabua** [comment]
@chenjian2664 could you rebase and ensure all prior comments are addressed. Then I hope I can get some help from @Praveen2112 and others

**Comment by mosabua** [comment]
Thanks for the rebase @chenjian2664 .. hopefully @Praveen2112 and @hashhar can take a look soon.

**Comment by mosabua** [comment]
Looks like there are some failures .. can you please ensure that the build works locally on your workstation and push any further necessary changes.

**Comment by metalshanked** [comment]
@chenjian2664,@mosabua:- Is this connector still on ? The native MySQL connector does is not fully compatible with Starrocks (certain commands and performance). Would be great if this official SR connector can be added.

Thanks!

**Comment by metalshanked** [comment]
@chenjian2664 - Is this connector still on ? The native MySQL connector does is not fully compatible with Starrocks (certain commands and performance). Would be great if this official SR connector can be added.

File: docs/src/main/sphinx/connector/starrocks.md
Comment (bitsondatadev) [inline docs/src/main/sphinx/connector/starrocks.md:56]:
```suggestion
To connect to multiple StarRocks servers, add another properties file to `etc/catalog` with a different name, for example, `starrocks.properties` to add the catalog `starrocks` for the second StarRocks instance. 
```

File: docs/src/main/sphinx/connector/starrocks.md
Comment (bitsondatadev) [inline docs/src/main/sphinx/connector/starrocks.md:92]:
```suggestion
StarRocks supports these [create table properties](https://docs.starrocks.io/en-us/latest/sql-reference/sql-statements/data-definition/CREATE%20TABLE)_.
```

File: docs/src/main/sphinx/connector/starrocks.md
Comment (bitsondatadev) [inline docs/src/main/sphinx/connector/starrocks.md:115]:
```suggestion
This is a list of columns to be used as the table's distribution key, the column type can not be `FLOAT` or `DOUBLE`. If not specified,
```

File: docs/src/main/sphinx/connector/starrocks.md
Comment (bitsondatadev) [inline docs/src/main/sphinx/connector/starrocks.md:122]: StarRocks supports these data types `<https://docs.starrocks.io/en-us/latest/sql-reference/sql-statements/data-types>`_.

File: docs/src/main/sphinx/connector/starrocks.md
Comment (bitsondatadev) [inline docs/src/main/sphinx/connector/starrocks.md:171]:
```suggestion
The StarRocks connector provides a schema for every StarRocks database.
```

**Comment by konovenski** [comment]
Is there any plans on releasing this?

**Comment by mosabua** [comment]
We are interested to get this reviewed, merged and released. However, currently there have not been sufficient reviews and testing feedback from community members and especially also maintainers.

**Comment by hackeryang** [comment]
> We are interested to get this reviewed, merged and released. However, currently there have not been sufficient reviews and testing feedback from community members and especially also maintainers.
> 
> If anyone can help testing, especially people with more knowledge about StarRocks, that would be helpful

We have back ported relevant source codes and used it in our own Trino 423 environment based on this project, which is also originated from the PR now:  https://github.com/footprintanalytics/trino-starrrocks/tree/starrocks_connector  

It seems work, so personally i think that it can be merged, after the jdk 23 ut error is fixed.

**Review by hackeryang — APPROVED** [review]

**Comment by mosabua** [comment]
I am adding the stale-ignore label since this is still a valid PR and useful connector to potentially add.

**Comment by mosabua** [comment]
Turns out I cant reopen since the head repository is gone. I think you have to either create a new PR or somehow update this one yourself @chenjian2664

**Pull request closed without merging**
